app/models/concerns/unparsed_run.rb

```diff
@@ -13,154 +18,286 @@ def parse_into_db
+
+        history = RunHistory.new(
+          run_id:               id,
+          attempt_number:       attempt.index.to_i,
+          realtime_duration_ms: attempt.time.real_time&.total_seconds&.*(1000) || nil,
+          gametime_duration_ms: attempt.time.game_time&.total_seconds&.*(1000) || nil,
+          pause_duration_ms:    attempt.pause_time&.total_seconds&.*(1000) || 0,
+          started_at:           dates[0] && DateTime.parse(dates[0].to_rfc3339),
+          ended_at:             dates[1] && DateTime.parse(dates[1].to_rfc3339)
         )
-      end)
+
+        # See https://github.com/glacials/splits-io/pull/474/files#r241242051
+        dates.compact.each(&:dispose)
+
+        container << history
+        next if container.size < MAX_CONTAINER_SIZE
+
+        write_run_histories(container)
+        container = []
+      end
+
+      return if container.blank?
+
+      write_run_histories(container)
     end
 
-    def write_segments(parsed_segments)
-      segs = []
+    def write_run_histories(histories)
+      result = RunHistory.import(histories)
+      return result unless result.failed_instances.any?
+
+      raise UnparsableRun
+    end
 
-      parsed_segments.each do |parsed_segment|
-        segs << Segment.new(
+    def process_segments(run)
+      container = {}
+      previous_segment = OpenStruct.new(realtime_end_ms: 0, gametime_end_ms: 0)
+
+      (0...run.len).each do |i|
+        lsc_segment = run.segment(i)
+        segment = Segment.new(
           run_id:         id,
-          segment_number: parsed_segment[:segment_number],
-          name:           parsed_segment[:name],
-
-          realtime_start_ms:             parsed_segment[:realtime_start_ms], # starts can be 0 (for the first segment)
-          realtime_end_ms:               zero_to_nil(parsed_segment[:realtime_end_ms]),
-          realtime_duration_ms:          zero_to_nil(parsed_segment[:realtime_duration_ms]),
-          realtime_shortest_duration_ms: zero_to_nil(parsed_segment[:realtime_best_ms]),
-
-          realtime_skipped: parsed_segment[:realtime_skipped],
-          realtime_reduced: false,
-          realtime_gold:    parsed_segment[:realtime_gold],
-
-          gametime_start_ms:             parsed_segment[:gametime_start_ms], # starts can be 0 (for the first segment)
-          gametime_end_ms:               zero_to_nil(parsed_segment[:gametime_end_ms]),
-          gametime_duration_ms:          zero_to_nil(parsed_segment[:gametime_duration_ms]),
-          gametime_shortest_duration_ms: zero_to_nil(parsed_segment[:gametime_best_ms]),
-
-          gametime_skipped: parsed_segment[:gametime_skipped],
-          gametime_reduced: false,
-          gametime_gold:    parsed_segment[:gametime_gold]
+          segment_number: i,
+          name:           lsc_segment.name || '',
+          icon:           lsc_segment.icon.presence # do something with this
         )
+
+        %w[real game].each do |timing|
```

Comment:
This was originally used so that we don't have the exact same block of code repeated once for gametime and once for realtime, but it looks ugly and is hard to follow.  Do you have any good solutions for this?

Reply:
🤔 yeah I agree this is pretty hard to grok. For realtime/gametime stuff I tend to go with the "1, 2, n" philosophy of DRY code -- write it for realtime, then copy paste it for gametime. If we ever add a third (nth) timing mechanism for god knows what reason, we can worry about cleanly reusing code then.

Reply:
That's a good way of looking at it.  For now I'll put it back to repeating since with only 2 it's not that much.